File: src/Sheet.hs
module Sheet where

import Data.List

-- Function to sum all elements if bool is true
sumIf :: (Num a) => [a] -> Bool -> a
sumIf xs b
  | b = sum xs
  | otherwise = 0

-- Function to count target in list
nummercounter :: [Int] -> Int -> Int
nummercounter xs x = x * length (filter (== x) xs)

-- Funktion to check if any element occurs exactly 4 times in a list
pasch4checker :: (Eq a) => [a] -> Bool
pasch4checker xs = length (filter (== head xs) xs) == 4 || length (filter (== xs !! 1) xs) == 4

-- Function to check if any element occurs exactly 3 times in a list
pasch3checker :: (Eq a) => [a] -> Bool
pasch3checker xs = any (\x -> length (filter (== x) xs) == 3) xs

-- Function to check for a fullHouse
fullHouseCheck :: (Eq a, Ord a) => [a] -> Bool
fullHouseCheck xs = sort (map length grouped) == [2, 3]
 where
  grouped = group (sort xs)

-- Function to check if list is "grStr"
grStrCheck :: (Ord a, Num a) => [a] -> Bool
grStrCheck xs = genericLength (nub xs) == 5 && sort xs == [1, 2, 3, 4, 5] || sort xs == [2, 3, 4, 5, 6]

-- Function to check if list is "klStr"
klStrCheck :: (Num a, Ord a) => [a] -> Bool
klStrCheck xs =
  let noDup = sort $ nub xs
   in noDup == [1, 2, 3, 4] || noDup == [2, 3, 4, 5] || noDup == [3, 4, 5, 6]

-- Function to check if all elements are the same
kniffelcheck :: (Ord a) => [a] -> Bool
kniffelcheck [] = False
kniffelcheck xs = minimum xs == maximum xs

File: src/Sheet.ex
defmodule Sheet do
  @spec sumif(list(integer()), boolean()) :: integer()
  def sumif(xs, y) do
    if y do
      Enum.sum(xs)
    else
      -1
    end
  end

  @spec nummercounter(list(integer()), integer()) :: integer()
  def nummercounter(xs, y) do
    count =
      xs
      |> Enum.filter(fn x -> x == y end)
      |> length()

    count * y
  end

  @spec pasch4checker(list(integer())) :: boolean()
  def pasch4checker(xs) do
    length(Enum.filter(xs, &(&1 = hd(xs)))) >= 4 ||
      length(Enum.filter(xs, fn x -> x == Enum.at(xs, 1) end)) >= 4
  end

  @spec pasch3checker(list(integer())) :: boolean()
  def pasch3checker(xs) do
    Enum.any?(xs, fn x -> length(Enum.filter(xs, fn y -> y == x end)) >= 3 end)
  end

  @spec full_house_check(list(integer())) :: boolean()
  def full_house_check(hand) do
    hand
    |> Enum.sort()
    |> Enum.group_by(& &1)
    |> Map.values()
    |> Enum.map(&length/1)
    |> Enum.sort() ==
      [2, 3]
  end

  @spec grstrcheck(list(integer())) :: boolean()
  def grstrcheck(hand) do
    (length(Enum.uniq(hand)) == 5 && Enum.sort(hand) == [1, 2, 3, 4, 5]) ||
      Enum.sort(hand) == [2, 3, 4, 5, 6]
  end

  @spec klstrcheck(list(integer())) :: boolean()
  def klstrcheck(numbers) when length(numbers) == 5 do
    numbers
    |> Enum.sort()
    |> Enum.uniq()
    |> check_consecutive(4)
  end

  def klstrcheck(_), do: false

  @spec check_consecutive(list(integer()), integer()) :: boolean()
  defp check_consecutive(numbers, count) do
    numbers
    |> Enum.chunk_every(count, 1, :discard)
    |> Enum.any?(fn chunk ->
      chunk == Enum.to_list(Enum.min(chunk)..Enum.max(chunk))
    end)
  end

  @spec kniffelcheck(list(integer())) :: boolean()
  def kniffelcheck(numbers) when is_list(numbers) and length(numbers) == 5 do
    Enum.all?(tl(numbers), fn x -> x == hd(numbers) end)
  end

  def kniffelcheck(_), do: false
end
